rfc-123-github-actions-ci.md

```diff
@@ -0,0 +1,102 @@
+- Already in use across some of our production applications. Examples: [business volunteer form][business-volunteer-form], [find support form][find-support-form], [vulnerable people form][vulnerable-people-form]
+- In the future, we can make our internal GitHub Actions into libraries so we’re not duplicating “set up Rails apps” steps across 50 apps and can update them in one place (perhaps `alphagov/actions`?). We already do something similar with `alphagov/govuk-jenkinslib`. See [Homebrew/actions](https://github.com/Homebrew/actions) for an example.
+
+### Cons
+
+- We have less control over what software is on the runners. But most of the time we only need a database, a version of Ruby and some gems. And updating our Jenkins CI instances with new packages is a massive pain.
+- Another system to learn, as well as the replatforming work already ongoing.
+- Duplication of build parameters and steps between all our apps, as they’re all Rails apps. This can be mitigated if we use a shared actions library as proposed earlier.
+
+### Findings from some initial explorations into using GitHub Actions
+
+#### `push` vs `pull_request` build triggers
+
+We considered the following options:
+
+1. Running `on: [pull_request, push]`. This means that internal devs get feedback on their commits when they push them, for example if they've forgotten to run RuboCop locally. And also so that external contributors PRs get CI feedback on their changes. It does mean that for long-running test suites, we could waste compute power, time and the planet's precious natural resources running things twice that don't need to be run twice.
+
+2. Running `on: push`. Actions don't run [by default on forked PRs](https://github.com/alphagov/govuk-docker/pull/337) - it's a security feature that GitHub has implemented. And [when a PR is raised from a fork](https://github.com/alphagov/govuk-docker/pull/337) and the receiving repo only has a push trigger, the tests don't run. So we'd be back to the annoying state that is [our current process for forked PRs][merge-a-pull-request].
+
+3. Running `on: pull_request`. Functionally, we run the same tests across the board, whichever of these we choose. But this one has the advantage that it's consistent behaviour for internal and external contributors. But internally we'd have to put up with not having tests run on branches (apart from `master`, we should probably include the `master` branch all the time), which would mean changing developer workflows to run more tests locally (ie RuboCop, a common thing that gets forgotten about), or being less worried about "exposing to the world" our force-pushes to get things right on the PR. We would also probably benefit from opening PRs for long-running branches earlier, using the draft PR functionality, and this would encourage that. This seems like the best option to _Issy_, but please feel free to disagree.
```

Comment:
```suggestion
3. Running `on: pull_request`. This has the advantage that it's consistent behaviour for internal and external contributors. Internally we'd have to put up with not having tests run on branches, prior to raising a PR. This would mean changing developer workflows to run more tests locally, or being less worried about "exposing to the world" our force-pushes to get things right on the PR. (Irrespective of this decision, we should always build the `master` branch when merge commits are `push`ed.)
```